libcnb-data/src/buildpack/mod.rs

```diff
@@ -34,22 +40,111 @@ use serde::Deserialize;
 /// type = "BSD-3-Clause"
 ///
 /// [[stacks]]
-/// id = "io.buildpacks.stacks.bionic"
-/// mixins = ["yj", "yq"]
+/// id = "*"
+/// "#;
+///
+/// let buildpack_toml = toml::from_str::<BuildpackToml<Option<toml::value::Table>>>(toml_str).expect("buildpack.toml did not match a known type!");
+/// match buildpack_toml {
+///     BuildpackToml::Single(buildpack) => println!("Found buildpack: {}", buildpack.buildpack.id),
+///     BuildpackToml::Meta(buildpack) => println!("Found meta-buildpack: {}", buildpack.buildpack.id),
+/// };
+/// ```
+#[derive(Deserialize, Debug)]
+#[serde(untagged)]
+pub enum BuildpackToml<BM> {
+    Single(SingleBuildpack<BM>),
+    Meta(MetaBuildpack<BM>),
```

Comment:
I'm not yet sold on the naming of these types. I feel like this PR is a good opportunity to resolve #168, or at least finalize a naming for the types so we can change the variable names separately to match the type names.

Since at least `SingleBuildpack` is used standalone in buildpack implementations, I don't think it really conveys what it is. Granted, naming CNB related things is hard with all the overloaded terms in the spec. When I first read this in a context where `SingleBuildpack` is used standalone (later in this PR), I stopped for a moment, thinking "why does this function get a buildpack as an argument?".

If we'd want to keep the `BuildpackToml` naming for the composite type, I could see us using `SingleBuildpackToml` and `MetaBuildpackToml` for the variant structs.

In general, I think I don't like the `BuildpackToml` name so much and would prefer `BuildpackDescriptor` instead. `SingleBuildpackDescriptor` and `MetaBuildpackDescriptor` work fine for the variants IMO.

Reply:
Yeah I agree all of the names are suboptimal, and that the CNB spec doesn't make this easy.

I like the idea of `BuildpackDescriptor` / `SingleBuildpackDescriptor` / `MetaBuildpackDescriptor`.

The other type name that bugs me, is `libcnb_data::buildpack::Buildpack`, which is the `buildpack` table inside `buildpacks.toml`. My first thought was that we could name it `BuildpackMetadata`, but then we also have the `metadata` table which sounds too similar to that. Other options for that might be `BuildpackTable`, `BuildpackProperties` (but then what's the difference between "properties" and "metadata"?), `BuildpackDetails` etc.

I guess really `metadata` is "custom metadata" and the `buildpack` table is `BuildpackMetadata`?

Eugh

Reply:
I think I named it `BuildpackToml` so you can blame me for that. I agree that `BuildpackDescriptor` is the better name.

Reply:
`libcnb_data::buildpack::Buildpack` is something for another PR I guess?

Reply:
Types renamed in ed15a71.